### ISSUE TITLE
in VideoFrame-from-VideoFrame constructor, allow setting visibleRect and displayWidth/displayHeight

### ISSUE THREAD

**Issue by dogben** [w3c/webcodecs#281]
It would be useful to be able to modify the visibleRect, displayWidth, and displayHeight attributes on a VideoFrame in order to represent a subset of the original frame or a different scale. Since VideoFrame is immutable, the best approach seems to be to add these parameters to the "copy constructor."

Example:
```
const subset = new VideoFrame(origVideoFrame, {
  visibleRect: {
    left: origVideoFrame.visibleRect.left + 100,
    top: origVideoFrame.visibleRect.top + 100,
    width: origVideoFrame.visibleRect.width - 200,
    height: origVideoFrame.visibleRect.height - 200
  },
  displayWidth: origVideoFrame.displayWidth,
  displayHeight: origVideoFrame.displayHeight,
});
```

The new parameters would behave identically to those in VideoFramePlaneInit. (However, @sandersdan seemed unsure if that would be the correct behavior if visibleRect is specified and displayWidth/displayHeight is not specified.)

For simplicity, [_Initialize Frame With Resource and Size_](https://w3c.github.io/webcodecs/#videoframe-initialize-frame-with-resource-and-size) should throw TypeError if these new parameters are specified.

This API would be very useful in combination with [MediaStreamTrack Insertable Media Processing using Streams](https://w3c.github.io/mediacapture-transform/) in order to produce a cropped or scaled version of an existing video track.

The alternative to this API for visibleRect would be to use createImageBitmap with cropping parameters, then create a new VideoFrame from the image bitmap; however, this would require conversions and copies, so the result would be lower-quality and slower. The alternative for displayWidth/displayHeight would be to draw the video frame to a canvas, which has the same drawbacks.

I expect this API to be very easy to implement.

**Comment by sandersdan** [comment]
> However, @sandersdan seemed unsure if that would be the correct behavior if visibleRect is specified and displayWidth/displayHeight is not specified.

To elaborate, computing the new displayWidth/displayHeight is not completely trivial, so it may make sense for the constructor to assist. It can take the new visible size and the source frame's pixel aspect ratio to compute a new display size.

In some cases the source frame may have a more precise pixel aspect ratio available than can be computed from the exposed properties, since those have been rounded to integers. (This may be a reason to pursue exposing pixel aspect ratio on frames.)

This is directly at odds with being a simple metadata override API, of course.

Currently I believe that the trade-off favors implementing the assistance, it seems unlikely that an app wouldn't want that feature by default.

**Comment by sandersdan** [comment]
Quick summary of the proposal above:

- Use override of visible rect as an efficient way to communicate crop to encoders.
- Use the constructor API to do this like for other metadata (timestamp, color space).
- Compute display size in the constructor using the pixel aspect of the source frame (unless display size is explicitly overridden). I think this is the behavior almost all uses would want.

Alternatives for a crop API:

- We could have an explicit VideoFrame.crop(). That's also something we can add later if there is demand for a more fluent API.
- We could put cropping on the encode API. This won't work for WebRTC Encoded Transform users since they don't use our encode API.
- We could distinguish crop from visible rect changes, with crop reducing the coded size. This isn't much different from cropping via ImageBitmap, and can't be optimized as readily.

Alternatives for display size behavior:

- We could continue to copy display size from the source frame even when the visible rect changes. If apps want to maintain pixel aspect, they would need to compute and override the display size explicitly.
- We could require display size to be supplied if visible rect is. (Basically punt on determining behavior.)

**Comment by fippo** [comment]
let me try to emphasize the importance of the use-case: cropping (maybe with a bit of zoom) is important for "smart framing". Group video calls in grid layout *may* get a bit more natural if everyone is shown at the same size, no matter how far away from the camera they are. That requires crop and a bit of zoom on the client side. I'd expect that use-case to be common enough that its worth optimizing.

canvas works but... feels clunky. The "copy constructor" seems elegant

**Comment by fippo** [comment]
@chcunningham do you have a chrome bug # for tracking implementation? :tada:

**Comment by chcunningham** [comment]
@fippo, I do! https://bugs.chromium.org/p/chromium/issues/detail?id=1243162

This feature is implemented, shipping in Chrome  95.0.4628.0 or higher (currently in DEV channel).
